Kick-off of API Blueprint module for apigility

miiila opened the pull request.

weierophinney commented on circle.yml:
```diff
@@ -0,0 +1,3 @@
+machine:
+  php:
+    version: 5.3.25
```
What is this file for? (I'm assuming it's for a CI service of some sort, but I don't know which one).

Also, if this is for a CI service, we typically test from our minimum supported version (currently 5.3, but we'll be bumping to 5.5 soon) through most stable, plus HHVM and PHP 7 nightlies; is the service this is for capable of testing multiple versions? If so, please set that up.

weierophinney replied:
Also, if it *is* for a CI service, let me know so I can setup the integrations with this repo. :)

weierophinney commented on composer.json:
```diff
@@ -0,0 +1,50 @@
+        "zendframework/zend-eventmanager": "~2.3",
+        "zendframework/zend-mvc": "~2.3",
+        "zendframework/zend-servicemanager": "~2.3",
+        "zendframework/zend-view": "~2.3",
```
We should specify `~2.4` at this point (we made that switch with v1.1).

weierophinney commented on composer.json:
```diff
@@ -0,0 +1,50 @@
+    "autoload": {
+        "psr-4": {
+            "ZF\\Apigility\\Documentation\\ApiBlueprint\\": "src/",
+            "ZFTest\\Apigility\\Documentation\\ApiBlueprint\\": "test/"
```
Move this one to an `autoload-dev` section.

weierophinney commented on config/module.config.php:
```diff
@@ -0,0 +1,86 @@
+            'ZF\Apigility\Documentation\ApiBlueprint\ApiBlueprintViewStrategy' => 'ZF\Apigility\Documentation\ApiBlueprint\ApiBlueprintViewStrategyFactory',
+        ),
+        'services' => array(
+            'ZF\Apigility\Documentation\ApiBlueprint\ApiBlueprintViewRenderer' => new ZF\Apigility\Documentation\ApiBlueprint\ApiBlueprintRenderer(),
```
In order to allow caching of configuration, this should be done as an invokable instead.

weierophinney commented on config/module.config.php:
```diff
@@ -0,0 +1,86 @@
+        ),
+        'accept_whitelist' => array(
+            'ZF\Apigility\Documentation\ApiBlueprint\Controller' => array(
+                0 => 'text/vnd.apiblueprint+markdown',
```
You can remove the index here.

weierophinney commented on phpunit.xml:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<phpunit
+	bootstrap="./test/Bootstrap.php">
```
Let's make this simpler: have it load "./vendor/autoload.php", and then remove `test/Bootstrap.php`. All the bootstrap is doing is setting up autoloading, so we really don't need it. (I'm going to be updating the various other modules to do this as well; it's a remnant from old ZF2 practices, which we recently removed with the component split.)

weierophinney commented on src/Action.php:
```diff
@@ -0,0 +1,91 @@
+<?php 
+/**
+ * @license   http://opensource.org/licenses/BSD-3-Clause BSD-3-Clause
+ * @copyright Copyright (c) 2014 Zend Technologies USA Inc. (http://www.zend.com)
```
2015 instead of 2014.

weierophinney commented on src/Action.php:
```diff
@@ -0,0 +1,91 @@
+
+use ZF\Apigility\Documentation\Operation as BaseOperation;
+
+class Action {
```
We follow [PSR-1](http://www.php-fig.org/psr/psr-1/) and [PSR-2](http://www.php-fig.org/psr/psr-2/); opening brace goes on the following line. I can fix these sorts of CS issues when I merge, though. (I'll also add a dependency on php-codesniffer when I do, which will provide the tools for detecting and fixing these.)

weierophinney commented on src/Action.php:
```diff
@@ -0,0 +1,91 @@
+	/**
+	 * @var BaseOperation
+	 */
+	private $operation;
```
4 spaces instead of tabs (for PSR1/2). Again, I can fix these during merge.

weierophinney commented on src/Action.php:
```diff
@@ -0,0 +1,91 @@
+	/**
+	 * @return boolean
+	 */
+	public function isEntityChanging() {
```
Maybe `allowsChangingEntity()`? Since it's a boolean, a natural phrase that implies a yes/no is nice for such methods:

```php
if ($action->allowsChangingEntity()) {
    ...
}
```

weierophinney commented on src/Api.php:
```diff
@@ -0,0 +1,51 @@
+<?php
+/**
+ * @license   http://opensource.org/licenses/BSD-3-Clause BSD-3-Clause
+ * @copyright Copyright (c) 2014 Zend Technologies USA Inc. (http://www.zend.com)
```
2015

weierophinney commented on src/Api.php:
```diff
@@ -0,0 +1,51 @@
+    private $api;
+
+    /**
+     * @var Blueprint\ResourceGroup[] 
```
Should be just `ResourceGroup[]`, as that class is in the same namespace.

weierophinney commented on src/Api.php:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    /**
+     * @return Blueprint\ResourceGroup[]
```
Same comment here.

weierophinney commented on src/ApiBlueprintModel.php:
```diff
@@ -0,0 +1,156 @@
+    public function getFormattedApiBlueprint() {
+        $model = new Api($this->variables['documentation']);
+        $this->apiBlueprint = 'FORMAT: ' . self::FORMAT . PHP_EOL;
+        $this->apiBlueprint .= 'HOST: ' . $_SERVER["REQUEST_SCHEME"] . "://" . $_SERVER["HTTP_HOST"] . self::$EMPTY_ROW; 
```
Don't use superglobals within classes. If you need information from superglobals, inject them via the constructor, or via argument parameters. In this particular case, the information here is present in the request instance, which you could either pass to the method, or pass the components you need as arguments directly.

weierophinney commented on src/ApiBlueprintModel.php:
```diff
@@ -0,0 +1,156 @@
+        parent::__construct();
+
+        // PHP < 5.6 is unable to concatenate constants into new constant
+        self::$EMPTY_ROW = PHP_EOL . PHP_EOL;
```
I would argue that, because this is intended for consumption via a browser or automated tooling, we can use `\n` instead of `PHP_EOL` throughout.

weierophinney replied:
Also, if we use "\n\n", then `EMPTY_ROW` can be a constant. :)

weierophinney commented on src/ApiBlueprintModel.php:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    /**
+     * @param Blueprint\ResourceGroup[] $resourceGroups
```
Should be `ResourceGroup[]`.

weierophinney commented on src/ApiBlueprintModel.php:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    /**
+     * @param Blueprint\Resource[] $resources
```
Should be `Resource[]`.

weierophinney commented on src/ApiBlueprintModel.php:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    /**
+     * @param Blueprint\Action[] $resources
```
Should be `Action[]`.

weierophinney commented on src/ApiBlueprintModel.php:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    /**
+     * @param Blueprint\Action $action
```
Should be `Action`.

weierophinney commented on src/ApiBlueprintModel.php:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    /**
+     * @var \Blueprint\Resource $resource
```
Should be `Resource`.

weierophinney commented on src/ApiBlueprintModel.php:
```diff
@@ -0,0 +1,156 @@
+     * @return string
+     */
+    private function getFormattedCodeBlock($codeBlock) {
+        return self::CODE_BLOCK_INDENT . \str_replace("\n", "\n" . self::CODE_BLOCK_INDENT, $codeBlock);
```
No need to prefix global functions with the `\` qualifier.

weierophinney commented on src/ApiBlueprintRenderer.php:
```diff
@@ -0,0 +1,38 @@
+<?php
+/**
+ * @license   http://opensource.org/licenses/BSD-3-Clause BSD-3-Clause
+ * @copyright Copyright (c) 2014 Zend Technologies USA Inc. (http://www.zend.com)
```
2015

weierophinney commented on src/ApiBlueprintRenderer.php:
```diff
@@ -0,0 +1,38 @@
+     * @return string The script output.
+     */
+    public function render($nameOrModel, $values = null) {
+    	echo $nameOrModel->getFormattedApiBlueprint();
```
Replace the `echo` with `return`; renderers should return the output, not echo it directly.

weierophinney commented on src/ApiBlueprintViewStrategy.php:
```diff
@@ -0,0 +1,76 @@
+
+        $headers = $response->getHeaders();
+        $headers->addHeaderLine('Content-Type', 'text/vnd.apiblueprint+markdown');
+    }
```
Once you update the renderer to return the results of rendering, you can update this to add:

```php
$response->setContent($e->getResult());
```

weierophinney commented on src/Resource.php:
```diff
@@ -0,0 +1,116 @@
+	private $uri;
+
+	/**
+	 * @var Blueprint\Action[]
```
Should be `Action[]`.

weierophinney commented on src/Resource.php:
```diff
@@ -0,0 +1,116 @@
+
+	/**
+	 * @param BaseService $service
+	 * @param BaseOperations[] $operations
```
Should be `BaseOperation[]`.

weierophinney commented on src/Resource.php:
```diff
@@ -0,0 +1,116 @@
+            $this->uri .= "?page={page}";
+        }
+
+    }
```
Mixture of spaces and tabs above; I'll fix this with phpcs later.

weierophinney commented on src/Resource.php:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    /**
+     * @return Blueprint\Action[] 
```
Should be `Action[]`

weierophinney commented on src/ResourceGroup.php:
```diff
@@ -0,0 +1,77 @@
+    protected $service;
+
+    /**
+     * @var Blueprint\Resource[]
```
Should be `Resource[]`.

weierophinney commented on test/ApiTest.php:
```diff
@@ -0,0 +1,29 @@
+
+class ApiTest extends TestCase
+{
+public function setUp () {
```
Indentation; I can fix during merge.

weierophinney commented:
@miiila Awesome! The only real actionable item I have for you is the question about `circle.yml`; I can take care of the other issues during merge, and I'll setup phpcs, Travis-CI, etc at that point as well.

Thanks!

miiila commented:
@weierophinney Thank for your nice and precise review! I addressed all issues (I hope :-)) except those you can easily fix during merge. I also removed circle.yml file (configuration for Circle CI), because setting up CI on Travis will be better, IMHO.

Please let me know if you will find anything else that would require my attention. Ale keep me updated what will happen now :-)
